### PR TITLE
feat(claude): set up Claude without a terminal — managed CLI install + in-app login

### LINKED ISSUE #271: Set up Claude without a terminal: managed CLI install + in-app login
Setting up the Claude CLI is a terminal task, so non-technical users can't finish it.

## The problem

Mila's AI features (title suggestion, automatic summary, Send action, Live AI) all
shell out to a `claude` binary that Mila does not install. Today a user who picks
Settings → AI Provider → Tool → "Claude (claude CLI)" has to leave the app and do
three things in a terminal before anything works:

1. **Install the CLI.** `curl -fsSL https://claude.ai/install.sh | bash`, or an
   `npm install -g` — both of which assume the user knows what a shell is, and the
   npm route additionally assumes a working Node install.
2. **Log in.** `claude` then has to be run interactively once to authenticate.
3. **Make Mila find it.** A GUI app launched from Finder inherits a stripped-down
   `PATH` from launchd, so a CLI that works in Terminal is frequently invisible to
   Mila. #196 widened the search (npm prefix, nvm/Volta/asdf, `~/.local/bin`) and
   there is an "Executable" override field, but both are mitigations for a setup
   step we never owned in the first place.

Every one of those steps is a place a non-technical user stops. The failure mode
is not even loud: the user selects the tool, the picker accepts it, and the first
title suggestion comes back as "Could not find claude on PATH".

## What it should be instead

One button in Settings → AI Provider — **Set up Claude** — that:

* downloads Anthropic's official native `claude` binary (the distribution
  `https://claude.ai/install.sh` serves) into a Mila-managed directory, verifying
  the publisher's SHA-256 and the code signature before it is ever executed;
* walks the user through the login inside the app: Mila runs `claude setup-token`,
  detects the authorization URL it prints, opens the browser, takes the pasted
  code in a sheet, and captures the long-lived OAuth token;
* stores that token in the Keychain and injects it as `CLAUDE_CODE_OAUTH_TOKEN`
  when — and only when — Mila runs its own managed binary, so anyone with a
  system-wide `claude` and their own login is untouched;
* shows an honest status line (Not set up / Installed, not signed in / Signed in)
  with a Test button and a Sign out affordance.

## Notes / constraints

* The managed binary must be verified before first execution: publisher checksum
  from the release manifest, `codesign` validity, and a TeamIdentifier match
  against Anthropic's Developer ID team. Anything that fails verification must be
  refused with a readable error rather than run.
* `claude setup-token` is an interactive TUI and produces no output at all when
  stdin is not a terminal, so the guided login needs a pseudo-terminal.
* The token is a credential: Keychain only, never `UserDefaults`, and it must not
  reach the unified log — including via the subprocess output that carries it
  (the log-hygiene rules from #193/#218 apply directly).
* Subprocess handling must follow `.claude/rules/python-subprocess.md` and
  `BlockingWork` (#251): `terminationHandler` + semaphore installed before
  `run()`, blocking waits on dedicated threads, every wait bounded.

### PR DESCRIPTION
**Pull request by urisland**
Closes #271

## What & why

Setting up Claude currently means a terminal: pipe an install script to a shell, run `claude setup-token`, then hope Mila's `PATH` search finds the result. For a large share of Mila's users that is where the feature ends — and it ends quietly, with the first title suggestion coming back as "Could not find claude on PATH".

This adds **Settings → AI Provider → Set up Claude**: one button that installs Anthropic's official CLI, signs the user in through their browser, and reports honestly what state the setup is in.

**Install.** The release resolution replicates `https://claude.ai/install.sh` rather than piping it to a shell — `GET /latest`, `/<version>/manifest.json` for the per-platform SHA-256 and size, `/<version>/<platform>/claude` — over URLSession, HTTPS to `downloads.claude.ai` only, with off-host redirects refused rather than followed. I verified this derivation against the real installer script while working on it: base URL, path layout, the `.platforms["<platform>"].checksum`/`.size` keys, the `darwin-arm64`/`darwin-x64` naming, the Rosetta `proc_translated` override, and the version-shape regex all match.

**Verify before executing.** Download → size → checksum → signature → install, and staging is deliberately *not* on the path `LLMRunner` resolves, so there is no window where a half-verified binary is reachable by anything that would run it. The checksum and the signature are not redundant: the checksum catches a truncated transfer, the signature keeps proving authorship if the manifest and binary come from the same compromised host. The requirement is anchored (`anchor apple generic` + team ID + signing identifier) — comparing identity strings read back from an unanchored signature would compare attacker-chosen values. Failures refuse with a readable reason and offer no "install anyway".

**Login.** Driven over a pty, because `claude setup-token` emits nothing at all on a pipe (it is an Ink TUI gating its render on `isatty(0)`). The parser handles what that implies: ANSI escapes, hard wrapping, and words positioned with cursor moves instead of spaces. The authorization URL is checked against an allowlist of Anthropic hosts before `NSWorkspace` opens it. The pty transport follows `.claude/rules/python-subprocess.md` and #251 — `terminationHandler` before `run()`, blocking reader/writer on dedicated threads, bounded SIGTERM→SIGKILL escalation, no unbounded wait.

**Token.** Keychain only, with verified writes and deletes. It never reaches UserDefaults, a log line, or a UI string: `ClaudeSetupState.logToken` drops every payload, and failure text built from CLI output is redacted first. The env is never logged and `LLMTestResult.command` is only path + args, so the injected token cannot reach the test panel either.

**Resolution + injection.** Order is override → managed install → `$PATH` search (#196). The managed binary beats the search because it is the one Mila can make promises about; the explicit Settings override still beats both, since that field is the documented escape hatch. The token is injected in exactly one place under one rule — only when the executable being launched *is* the managed binary — so a user's own `claude` with its own login (possibly a different account, possibly corporate SSO) is untouched, and nothing is ever removed from the environment.

## Bugs found in the pre-existing work-in-progress

This finishes an interrupted implementation. Three real defects in the inherited code, each now pinned by a test:

1. **A terminal state could be overwritten within one chunk.** `ingest` checked `finished` once before its event loop instead of per event. A chunk carrying both a fatal rejection and a token-shaped string ran the token event afterwards — turning a reported failure into `.ready` and handing back a credential from a run just called broken.
2. **The login child inherited the token it was about to replace.** `PTYProcess` defaulted to the normal child environment, so `setup-token` ran holding the existing credential. A "sign in again" that silently returns the OLD token is the failure hardest to notice: the UI says Signed in and the Test passes. It now runs with no credential at all.
3. **The session cleared its own only strong reference from inside its `didSet`.** It survived only because every caller happened to reach it via optional chaining on a weak `self` — a property of the call sites, not of the code. Now released on the next main-actor turn, guarded on identity so "Try again" doesn't tear down the sign-in it just started.
4. **A token split across a pty read boundary was reported truncated.** Found by CI on the first run of the new parser suite — the assertion I expected to be trivially true was not. A *prefix* of a token is itself token-shaped (`sk-ant-` plus sixteen more characters), so a chunk boundary mid-token made the match succeed on the half that had arrived: the flow stored a truncated credential and reported success, and the failure would have surfaced much later as an auth error with the Settings row still saying "Signed in". A match running to the end of the buffer is now refused while the stream is live (the match is greedy, so ending earlier proves the value is complete), with a last-chance scan at child exit so a token that genuinely is the final bytes is still recovered — exactly once.

Also fixed a compile break: a bare `/regex/` literal, which needs the `BareSlashRegexLiterals` upcoming-feature flag that `project.yml` does not set at Swift 5.10. Switched to the `#/…/#` extended-delimiter form.

## How it was tested

**Verified.** `make build` — BUILD SUCCEEDED. `make package-test` — passed, 101 MilaKit tests, 0 failures. **CI is green on this branch**: `unit-and-ui-tests`, `build-and-test`, `e2e`, `diarize`, `remote-transcription-e2e`, both mock meeting jobs, `UI layout tests` and `Linked issue` all pass. The first CI run failed on one of the new tests, which is how bug 4 above was found; the fix and its regression tests are in the final commit and the suite is green.

Five new suites, none of which touch the network or execute a real `claude`: release-URL and version-guard rules, manifest decoding, executable-resolution precedence, the token-injection rule (including that the Keychain is not read for a non-managed binary), the install ordering and every refusal path (checksum, size, wrong publisher, broken signature, junk version), the parser over realistic escape-laden transcripts, the state machine end to end (URL detection, paste, capture, rejection-and-retry, exit-0-without-token, launch failure, cancellation, and both timeouts via an injectable seam), the persisted verify-status rules, sign-out verification, and a walk of the whole defaults suite asserting no value contains a token.

**Not verified — needs manual checking on a real Mac.** I could not exercise the parts that require a real credential, and deliberately did not run `claude setup-token` (it opens a browser on the developer's machine). Running `make test` locally was ruled out, so CI is where the new suites execute — they now pass there.

### Needs manual verification

1. **Button → install.** Settings → AI Provider → Tool → "Claude (claude CLI)". With no managed install, the row should read *Not set up*. Press **Set up Claude**; a progress bar should run and the binary should appear at `~/Library/Application Support/Mila/bin/claude`.
2. **Signature gate actually refuses.** Replace that file with any other signed binary (or an ad-hoc-signed one) and press **Reinstall**… then confirm a genuine tampered download is refused — easiest check is that a corrupted staged file produces the checksum error rather than an install.
3. **Real browser login.** The sheet should appear, the browser should open Anthropic's page exactly once (not twice), and pasting the code should move the row to *Signed in*. Try pasting a deliberately wrong code first: it should show the CLI's own rejection and let you retry without restarting.
4. **Token round-trip.** Confirm a `claudeSetup.oauthToken` item exists in Keychain Access, and that `defaults read io.island.whisper.IslandWhisper` shows no token anywhere.
5. **Test button.** Press **Test**; it should reach *Signed in and working*, and survive a Settings reopen (persisted verification).
6. **A system `claude` is untouched.** With your own `claude` on `PATH` and its own login, confirm it still works and never receives `CLAUDE_CODE_OAUTH_TOKEN`.
7. **Sign out.** Should remove the Keychain item and flip to *Installed, not signed in*; "sign out and delete the binary" should return the row to *Not set up*.

## Open risks

- **The success transcript is unobserved.** The token pattern is derived from the `sk-ant-oat…` prefix the shipped binary carries, not from a captured successful run. If a future CLI version stops printing the token, the flow fails loudly ("finished but didn't print a token Mila could read") rather than claiming success — but it does fail, and item 3 above is the only way to confirm it currently works.
- **The parser is coupled to a TUI's layout.** Markers are matched whitespace-insensitively and wrapped lines are rejoined precisely because that layout is not a contract, but a sufficiently different future `setup-token` UI would need re-pinning.
- **Compressed artifacts.** The official installer can serve `claude.zst`; Mila fetches the uncompressed `claude`, which exists today. If a future release ships only the compressed form, install would need to handle it.
- **No auto-update for the managed binary.** Deliberate for v1 — "Reinstall" is the recovery path, since the failure to cover is "my CLI stopped working", not "my CLI is three days old".

## Checklist

- [x] Builds locally (`make build`); `make package-test` passes; `make test` was not run locally (per the constraints this work was done under) — CI's `unit-and-ui-tests` job runs it and is green.
- [x] No secrets, credentials, tokens, or internal infrastructure references added
- [x] Follows the conventions in `CLAUDE.md`
- [ ] User-facing change is noted in the README changelog — not added; this is not a release bump, and `RELEASE_NOTES/` is written at version-bump time

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added managed Claude CLI installation with download validation, progress reporting, and secure replacement of existing installations.
  - Added Claude setup controls in Settings for installation, guided sign-in, testing, reinstalling, and signing out.
  - Added support for detecting and using an existing system-installed Claude CLI.
  - Added secure Keychain storage for Claude credentials.
  - Claude launches now prefer Mila’s managed installation and use its stored credential.

- **Bug Fixes**
  - Improved sign-in prompts, authorization links, cancellations, timeouts, credential redaction, and overlapping installation handling.

---

> [!NOTE]
> **High Risk**
> Downloads and executes a third-party binary, stores long-lived OAuth credentials in Keychain, and changes which `claude` process runs with injected secrets—mitigated by pinned hosts, signature requirements, and path-scoped token injection.
> 
> **Overview**
> Adds **Settings → AI Provider → Set up Claude**: one flow to download Anthropic’s official `claude` binary, verify it, sign in via browser, and persist the OAuth token—without asking users to use Terminal.
> 
> **Install pipeline** mirrors `install.sh` against `downloads.claude.ai` (version → manifest → binary), streams the ~200 MB artifact to a **staging** path, then enforces **size → SHA-256 → Developer ID signature** before anything lands at `~/Library/Application Support/Mila/bin/claude`. Off-host redirects are refused; failed installs remove staged bytes.
> 
> **Login** runs `claude setup-token` over a **pty** (the CLI is silent on pipes). A parser strips TUI escapes, dewraps URLs, allowlists authorization hosts, and redacts tokens from transcripts; `ClaudeSetupTokenSession` drives browser open, code paste, timeouts, and cancellation.
> 
> **Runtime wiring**: `LLMRunner.resolveExecutable` now prefers **override → managed `claude` → PATH**; `childEnvironment` injects `CLAUDE_CODE_OAUTH_TOKEN` **only** when the launched executable is the managed binary (Keychain read is lazy). `setup-token` explicitly runs **without** an inherited token so re-auth cannot silently reuse the old credential.
> 
> **UI/state**: `ClaudeSetupSettings` + `ClaudeSetupSection` cover install progress, sign-in sheet, Test/Reinstall/Sign out, system-CLI detection, and persisted “verified” status. Tokens live in **Keychain** via `KeychainClaudeTokenStore`; `KeychainHelper.isAbsent` verifies deletes.
> 
> Extensive new unit tests cover install refusals, resolution/injection rules, parser edge cases, and settings persistence—without network or a real `claude`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 62bfbb396e6e9fbcfbce9eff0efaa786ecfbe8da. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: Mila/Models/ClaudeSetupSettings.swift [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+    }
+
+    /// Abandon an in-progress sign-in.
+    func cancelSignIn() {
+        session?.cancel()
+        session = nil
+        switch state {
+        case .signingIn, .awaitingCode, .verifying: state = .idle
+        case .idle, .installing, .ready, .failed: break
+        }
```
[the remark this answers is not in the document]
Comment (urisland) [reply Mila/Models/ClaudeSetupSettings.swift:396]: Fixed in ed1dfc4: `cancelSignIn()` now cancels `installTask` and leaves `.installing` for `.idle`, and `performInstall` treats a cancelled URLSession task (`Task.isCancelled`) as a cancel rather than a failure — so the chained `startSignIn()` browser open no longer happens after a dismissed download.

File: Mila/Actions/ClaudeSetupTokenSession.swift [hunk 2]
```diff
@@ -0,0 +1,576 @@
...
+            }
+            lastRejection = ClaudeSetupTokenParser.redact(message)
+            state = .awaitingCode(authorizationURL)
+            arm(timeouts.authorization,
+                reason: "Sign-in timed out. Start setup again when you're ready.")
+
+        case .token(let token):
+            onToken?(token)
+            finish()
+            state = .ready
```
[the remark this answers is not in the document]
Comment (urisland) [reply Mila/Actions/ClaudeSetupTokenSession.swift:316]: Fixed in ed1dfc4: `onStateChange` keeps a `.failed` recorded by `store(_:)` instead of copying the session's trailing `.ready` over it — the Keychain-save error stays visible; the session is still released.

File: Mila/Actions/ClaudeBinaryInstaller.swift [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+    /// against a pinned requirement, which is a stricter statement than the
+    /// first-run Gatekeeper prompt makes; and that prompt is not available to
+    /// us anyway, because the binary is launched as a subprocess of a GUI app
+    /// where a refusal surfaces as an unexplained non-zero exit.
+    private func moveIntoPlace(_ staged: URL) throws -> URL {
+        let destination = ClaudeManagedInstall.binaryURL(appSupportRoot: appSupportRoot)
+        try createDirectory(destination.deletingLastPathComponent())
+        do {
+            if fileManager.fileExists(atPath: destination.path) {
+                _ = try fileManager.replaceItemAt(destination, withItemAt: staged)
```
[the remark this answers is not in the document]
Comment (urisland) [reply Mila/Actions/ClaudeBinaryInstaller.swift:227]: Fixed in d73fbea: `moveIntoPlace` re-asserts 0o755 on the final path after `replaceItemAt`/`moveItem`, and the reinstall test now writes a non-executable leftover and asserts `isExecutableFile` after install.

File: Mila/Actions/ClaudeBinaryInstaller.swift [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+                    } catch {
+                        continuation.resume(throwing: ClaudeInstallError.installFailed(
+                            error.localizedDescription))
+                    }
+                }
+                box.task = task
+                task.resume()
+            }
+        } onCancel: {
+            box.task?.cancel()
```
[the remark this answers is not in the document]
Comment (urisland) [reply Mila/Actions/ClaudeBinaryInstaller.swift]: Fixed in d73fbea with the proposed adopt/cancel lock pattern — a cancel that races the assignment now cancels the task instead of being dropped.

File: Mila/Actions/ClaudeSetupTokenParser.swift [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+            events.append(.awaitingCode)
+        }
+        // Rejections repeat: the CLI re-prompts rather than exiting, so the
+        // second bad paste has to be visible too. Counted rather than
+        // remembered as a flag.
+        let rejections = Self.rejectionMessages(in: text)
+        if rejections.count > rejectionsSeen {
+            events += rejections[rejectionsSeen...].map { Event.codeRejected($0) }
+            rejectionsSeen = rejections.count
+        }
```
[the remark this answers is not in the document]
Comment (urisland) [reply Mila/Actions/ClaudeSetupTokenParser.swift:160]: Fixed in d73fbea: on truncation `rejectionsSeen` is rebased to `min(rejectionsSeen, retained-in-window)`, so an evicted rejection can no longer suppress the next one.

File: Mila/Actions/ClaudeSetupTokenSession.swift [hunk 6]
```diff
@@ -0,0 +1,576 @@
...
+        // restoring the terminal can take a moment) it gets killed.
+        BlockingWork.onDedicatedThread(named: "io.island.mila.claude-setup.reap") { [weak self] in
+            if self?.exited.wait(timeout: .now() + 2) == .timedOut {
+                kill(process.processIdentifier, SIGKILL)
+                _ = self?.exited.wait(timeout: .now() + 3)
+            }
+            // Closing the master unblocks the reader thread's `read` even if a
+            // grandchild inherited the slave and is holding it open.
+            self?.closeMaster()
+        }
```
[the remark this answers is not in the document]
Comment (urisland) [reply Mila/Actions/ClaudeSetupTokenSession.swift:576]: Skipping deliberately: a reader parked on a dedicated thread that may leak if a grandchild holds the slave open is this repo's documented tradeoff (.claude/rules/python-subprocess.md — a leaked dedicated thread costs one thread, not a slot in a shared pool). `BlockingWork.onDedicatedThread` exists precisely so we do not build poll/kqueue cancellation machinery around every blocking read; the bounded waits around it keep the flow itself from hanging.

File: Mila/Models/ClaudeTokenStore.swift [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        // its `OSStatus`, and a keychain write can fail for reasons that have
+        // nothing to do with this app (a locked keychain, a duplicate item left
+        // by an older build).
+        return load() == trimmed
+    }
+
+    @discardableResult
+    func delete() -> Bool {
+        KeychainHelper.delete(key: key)
+        return load() == nil
```
[the remark this answers is not in the document]
Comment (urisland) [reply Mila/Models/ClaudeTokenStore.swift]: Fixed in d73fbea: added `KeychainHelper.isAbsent(key:)` (positive `errSecItemNotFound` proof) and `delete()` verifies through it instead of `load() == nil`, so a keychain read error no longer reads as a successful delete.

File: Mila/Views/ClaudeSetupSection.swift [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+
+            case .signedIn:
+                testButton
+                reinstallButton
+                Button("Sign out…") { showRemoveConfirmation = true }
+                    .accessibilityIdentifier("ai.provider.claudeSetup.signOut")
+
+            case .working:
+                Button("Cancel") { setup.cancelSignIn() }
+                    .accessibilityIdentifier("ai.provider.claudeSetup.cancel")
```
[the remark this answers is not in the document]
Comment (urisland) [reply Mila/Views/ClaudeSetupSection.swift]: Addressed in ed1dfc4: Cancel now cancels the active install (see the Bugbot thread above). For Test: it stays non-cancellable by design — it is bounded at 90s — but the UI no longer lies about it: while testing, the row shows the spinner-labelled Test button instead of a Cancel that would do nothing.

**Comment by urisland** [comment]
Manual verification completed by the maintainer on the real flow: managed download + signature check, browser login, code paste, token stored, and the Test button — all worked end-to-end on a machine with no `claude` on PATH. The "Needs manual verification" items in the description are confirmed.

bugbot run

File: Mila/Actions/ClaudeSetupTokenParser.swift [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+        buffer += chunk
+        if buffer.count > Self.bufferLimit {
+            buffer = String(buffer.suffix(Self.bufferLimit))
+            // Truncation can evict already-reported rejections from the
+            // window; `rejectionsSeen` must shrink with them or the count
+            // stays ahead of what is visible and the NEXT rejection gets
+            // swallowed (`rejections.count > rejectionsSeen` never fires).
+            let retained = Self.rejectionMessages(
+                in: Self.dewrap(Self.sanitize(buffer), width: width)).count
+            rejectionsSeen = min(rejectionsSeen, retained)
```
[the remark this answers is not in the document]
Comment (urisland) [reply Mila/Actions/ClaudeSetupTokenParser.swift]: Confirmed and fixed in 4494a29: the clamp is replaced by an eviction delta — rejectionsSeen shrinks by (count before truncation − count after), so a rejection arriving in the overflowing chunk itself is still emitted. Added a regression test reproducing exactly this scenario (first rejection reported, 70 KiB filler + second rejection in one chunk).

File: Mila/Models/ClaudeSetupSettings.swift [hunk 10]
```diff
@@ -0,0 +1,492 @@
...
+    /// Cancelling `installTask` matters even mid-download: without it the
+    /// transfer keeps running, and `setUp()`'s task would then proceed to
+    /// `startSignIn()` and open a browser the user has already walked away
+    /// from.
+    func cancelSignIn() {
+        installTask?.cancel()
+        session?.cancel()
+        session = nil
+        switch state {
+        case .signingIn, .awaitingCode, .verifying, .installing: state = .idle
```
[the remark this answers is not in the document]
Comment (urisland) [reply Mila/Models/ClaudeSetupSettings.swift:394]: Fixed in 4494a29 with an install generation: setUp/reinstall/cancel each bump it, performInstall's state writes (initial, progress, success, both catch paths) and the chained startSignIn are all guarded on the generation still being current, so a cancelled task that resumes late writes nothing. No deterministic unit test for the interleaving — exercising it would need a blockable downloader seam at the settings level; the guard is unconditional rather than timing-dependent, which is the property that matters.

File: Mila/Models/ClaudeSetupSettings.swift [hunk 11]
```diff
@@ -0,0 +1,511 @@
...
+        state = .installing(progress: -1)
+        do {
+            let result = try await installer.install { progress in
+                Task { @MainActor [weak self] in
+                    guard let self, generation == self.installGeneration,
+                          case .installing = self.state else { return }
+                    self.state = .installing(progress: progress)
+                }
+            }
+            guard generation == installGeneration else { return false }
```
[the remark this answers is not in the document]
Comment (urisland) [reply Mila/Models/ClaudeSetupSettings.swift]: Confirmed and fixed in cc53547, from both ends: (1) the installer now runs Task.checkCancellation() after the download and again right before moveIntoPlace, so a late cancel stops the disk write in all but the final instants (its catch already removes the staged file); (2) for a success that does complete, the disk-mirroring bookkeeping — version default, refreshInstalledState, clearVerification — runs regardless of generation, and only the flow-owned writes (spinner state, chained sign-in) stay generation-guarded. Settings can no longer describe a machine that stopped existing.

File: Mila/Models/ClaudeSetupSettings.swift [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+            defaults.set(result.version, forKey: Keys.installedVersion)
+            refreshInstalledState()
+            // A new binary invalidates the proof: the Test that passed was
+            // against the *previous* one. `restoreVerification`'s parameter
+            // check would catch this on the next launch anyway; clearing it now
+            // means the UI doesn't claim a verification it no longer has.
+            clearVerification()
+            // Only the flow-owned writes stay generation-guarded: the spinner
+            // state and the chained sign-in belong to whoever runs NOW.
+            guard generation == installGeneration else { return false }
```
[the remark this answers is not in the document]
Comment (urisland) [reply Mila/Models/ClaudeSetupSettings.swift:295]: Fixed in 3744467 by serializing as suggested: setUp/reinstall capture the predecessor task, cancel it, and the replacement awaits its completion before touching anything — installer runs can no longer overlap, so the reverse-completion interleaving is gone structurally rather than probabilistically. The generation still guards flow-owned state. On the reverse-completion test: exercising it deterministically needs a blockable-downloader seam at the settings level that does not exist yet; the serialization is unconditional (an await, not a timing assumption), which is the property the test would pin.

File: Mila/Models/ClaudeSetupSettings.swift [hunk 13]
```diff
@@ -0,0 +1,530 @@
...
+        previous?.cancel()
+        installGeneration += 1
+        let generation = installGeneration
+        installTask = Task { @MainActor [weak self] in
+            // Serialize on the predecessor: a cancelled install can slip past
+            // its final cancellation check and still be writing the binary and
+            // its bookkeeping. Starting the next install only after the old
+            // one fully unwinds removes every reverse-completion interleaving
+            // — the generation alone orders state ownership, not disk writes.
+            await previous?.value
```
[the remark this answers is not in the document]
Comment (urisland) [reply Mila/Models/ClaudeSetupSettings.swift:233]: Confirmed and fixed in 62bfbb3: setUp/reinstall publish their busy state synchronously before queueing (so the row shows work in progress during the predecessor wait and the busy guard swallows further clicks), the chained startSignIn releases the placeholder before its own busy guard, and signOut now cancels + disowns any queued install via the generation bump — it can no longer put back a binary the user just removed.
